client/src/redux/sagas/buildSetSaga.js

```diff
@@ -1,16 +1,67 @@
-import { takeEvery, call, put } from 'redux-saga/effects';
-import { BUILD_DATA } from '../constants';
-import { fetchBuildSets } from '../api';
-import { setBuildSets, setBuildSetsError } from '../actions';
+import { takeEvery, call, put, select } from 'redux-saga/effects';
+import { COMMITS } from '../constants';
+import { fetchBuildSets, fetchBuildReq, fetchBuilds } from '../api';
+import { setBuildSetsError, setBuilds } from '../actions';
+
+const commitSha = state => state.commits;
+function buildStr(builds) {
+ return builds
+  .map(build => 'buildrequestid__contains=' + build.buildrequestid)
+  .join('&');
+}
+
+function getBuildSetReqString(commits, buildData) {
+ return commits
+  .map(commit =>
+   buildData.filter(bd => bd.sourcestamps[0].revision === commit.sha)
+  )
+  .flat()
+  .map(bd => 'buildsetid__contains=' + bd.bsid)
+  .join('&');
+}
+
 function* handleBuildsLoad() {
  try {
   const buildSets = yield call(fetchBuildSets);
-  yield put(setBuildSets(buildSets));
+  const commits = yield select(commitSha);
+  let buildReqStr = getBuildSetReqString(commits, buildSets);
+  if (buildReqStr.length > 0) {
+   var bsID = yield call(fetchBuildReq, buildReqStr);
+  }
+  let buildstr = buildStr(bsID);
+  if (buildstr.length > 0) {
+   let buildData = yield call(fetchBuilds, buildstr);
+   const filteredBS = commits.map(commit =>
+    buildSets.filter(bd => bd.sourcestamps[0].revision === commit.sha)
+   );
+   const shas = commits.map(commit => commit.sha);
+   const output = filteredBS.map((obj, index) => {
```

Comment:
Here you are iterating over `commits` two times, which is an overhead.
Remember my getBuildSetReqString refactoring and try to refactor this in functional style too :)